Initial fixes run through: prev/place order, type hinting and general formatting

Robert-Zacchigna opened the pull request:
@jessecooper This is my initial run through to fix the prev/place order. I added type hinting basically everywhere and fixed some formatting.

Another thing i noticed while i was testing things, if you try to preview/place an order of something that is more than the current value of available cash in your account, the API returns an error. As a result, line **52** of `order.py` raises the request error from the etrade API response BUT in doing so doesn't allow the user to see the etrade API error output, thus outputting the HTTP 400 bad response error instead.

This is what the error code would like from the etrade API:
```json
{
    "Error": {
        "code": "1021",
        "message": "We cannot accept this order because there are insufficient funds in your account."
    }
}
```

After looking through peoples issues, i almost guarantee that this is the main issue people were facing. The API for whatever reason won't even let you preview the order if you don't have enough funds in your account (happens with my work around functions as well). I've only commented it out for now as I'm not sure how you want to handle the error codes.

Lastly, untangling kwargs appears to be a bigger task than i thought, it would require a rewrite of quite a few things to work correctly with how things currently work. I haven't made those changes as i wanted to get your feedback on all of my work so far and see how you wanted to proceed.

I tried to modify the tests as well but admittedly i was never really good about unittesting, so I might need some help in that department.

xozxro commented:
This is good.

My company is running into this issue and finding this pull request is the only thing I found to figure it out.

Suggest that this is dealt with ASAP. Thanks for the great work.

jessecooper commented:
My apologies for the lag on this. Had to wrap up things in the day job before taking vacation. I am now getting back into the swing of things. 

> Lastly, untangling kwargs appears to be a bigger task than i thought, it would require a rewrite of quite a few things to work correctly with how things currently work. I haven't made those changes as i wanted to get your feedback on all of my work so far and see how you wanted to proceed.

For the kwargs. I am fine with not untangling all of that at this time. Some of the APIs have a lot of args and trying to convert and cover them all will take time. 

Starting this review on the GitHub app from my phone so I'll have to make this a multi post comment...

jessecooper commented:
> Another thing i noticed while i was testing things, if you try to preview/place an order of something that is more than the current value of available cash in your account, the API returns an error. As a result, line 52 of order.py raises the request error from the etrade API response BUT in doing so doesn't allow the user to see the etrade API error output, thus outputting the HTTP 400 bad response error instead.

Here instead of using the raise from status we should look at the status code and if it is greater than 200 raise a custom exception with the E-Trade error text and status code.

I will start the full code review tomorrow night.

Thanks you again for the PR and the patience with the comments and review.

jessecooper commented:
Ok it's a bit harder to review a change on the phone in the GitHub app. I'll be back at an actual screen come Monday but it looks good so far just want to give it a close look.

jessecooper commented on pyetrade/order.py:
```diff
@@ -10,86 +10,87 @@
 import jxmlease
+import xmltodict
+import dateutil.parser
+from requests import sessions
```
Is this import only used for the type hint? You might be able to do it off of `OAuth1Session` object.

Robert-Zacchigna replied:
indeed, changed it to `OAuth1Session.request` in the latest commit.

jessecooper commented:
@Robert-Zacchigna It looks good. The tests are only failing right now because the Union type was introduced in Python 3.10 but I am fine with deprecating support for older versions of python. 
https://docs.python.org/3/library/typing.html#typing.Union

I am willing to merge this even without the custom exception but I would like to put a `#TODO` comment on that if you do not want to create the custom exception. Otherwise it is just trying to removed the session import if we can avoid the import for type.

Robert-Zacchigna commented:
@jessecooper Hey, ya no problem at all, i was actually just about to post another message just in case you forgot (all good). 

Please review my latest commit:
- I updated the imports to reduce the overhead a bit, since for several of them its only using one function from the package.
- I remediated the Union issue (i think) and explicitly imported the Union type, didn't realize the shorthand was only recently added in 3.10
- Lastly, i updated the `get_request_result` function to handle the Etrade API errors, I'm simply checking if the `'Error'` key is present and then outputting the error code and message (let me know if you want something more refined):
       - Example Output (attempting to trade after hours): 
       - ```Exception: Etrade API Error - Code: 7029, Msg: The order you entered is not valid at this time. Please contact Customer Service at 1-800-387-2331 for assistance.```

jessecooper commented on pyetrade/order.py:
```diff
@@ -520,44 +543,37 @@ def place_changed_equity_order(self, resp_format=None, **kwargs):
+        return self.perform_request(self.session.put, api_url, payload, "xml")
 
-    def cancel_order(self, account_id, order_num, resp_format=None):
+    def cancel_order(self, account_id_key: str, order_num: int) -> dict:
```
We will need to add back the response format here
https://github.com/jessecooper/pyetrade/actions/runs/4473576965/jobs/7872705243

Robert-Zacchigna replied:
fixed, see latest commit

jessecooper commented on pyetrade/order.py:
```diff
@@ -10,86 +10,92 @@
+from typing import Union
+from jxmlease import emit_xml
 from requests_oauthlib import OAuth1Session
+from xmltodict import parse as xmltodict_parse
```
can we just import this in a similar way to:
https://github.com/jessecooper/pyetrade/blob/master/pyetrade/accounts.py#L10
I would like to retain the namespace. It also seems to not be working in the tests.
https://github.com/jessecooper/pyetrade/actions/runs/4473576965/jobs/7872705243#step:5:95

Robert-Zacchigna replied:
done, see latest commit

jessecooper commented on tests/test_order.py:
```diff
@@ -2,21 +2,21 @@
 """pyetrade authorization unit tests
```
the test for `test_place_equity_order` the xml assertion will need to be changed to match the `xmltodict` return.
https://github.com/jessecooper/pyetrade/actions/runs/4473576965/jobs/7872705243#step:5:114

example xml tests from accounts.py
https://github.com/jessecooper/pyetrade/blob/master/tests/test_accounts.py#L49

Robert-Zacchigna replied:
i think i did it correctly, unit testing was never my strong suit, so please let me know if i did it wrong.

Robert-Zacchigna replied:
ok fixed the unit-tests, i was able to test it all after setting up my local env, all tests pass now.

Robert-Zacchigna commented:
@jessecooper See latest commit, unit-tests should all pass now, took me a little bit to setup my local env to get test them all but should be good now.

Please double check that they are all correct.

jessecooper commented:
@Robert-Zacchigna Thank you for your PR and patients! It is very appreciated.

Robert-Zacchigna commented:
No problem at all, this has been learning experience for me as well.
